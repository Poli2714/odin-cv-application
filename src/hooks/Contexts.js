import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
