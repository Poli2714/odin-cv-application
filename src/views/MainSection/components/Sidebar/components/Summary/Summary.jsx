import { useContext } from 'react';

import { CharacterLimit } from 'src/components/ui';
import { Form, Label, TextArea } from 'src/components/forms';
import { InputWrapper } from 'src/components/layout';
import { SummaryContext } from 'src/hooks/Contexts';

const limit = 200;
const rows = 10;

function Summary() {
  const { summary, setSummary } = useContext(SummaryContext);
  const handleChange = e => setSummary(e.target.value);

  return (
    <Form>
      <InputWrapper>
        <Label labelName="Short summary" labelFor="summary" />
        <CharacterLimit value={summary} limit={limit}>
          <TextArea
            autofocus={true}
            name="summary"
            placeholder="Your professional summary"
            rows={rows}
            maxLength={limit}
            value={summary}
            onChange={handleChange}
          />
        </CharacterLimit>
      </InputWrapper>
    </Form>
  );
}

export default Summary;
