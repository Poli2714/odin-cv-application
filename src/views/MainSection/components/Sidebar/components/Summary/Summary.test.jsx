import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';

import Summary from './Summary';
import { SummaryContext } from 'src/hooks/Contexts';

vi.mock('src/components/layout', () => ({
  InputWrapper: vi.fn(({ children }) => <div>{children}</div>),
}));

vi.mock('src/components/forms', () => ({
  Form: vi.fn(({ children }) => (
    <form action="" data-testid="form" method="get">
      {children}
    </form>
  )),
  Label: vi.fn(() => <label htmlFor="summary">Mock label</label>),
  TextArea: vi.fn(({ value, onChange }) => (
    <textarea
      id="summary"
      rows={10}
      maxLength={200}
      value={value}
      onChange={onChange}
    ></textarea>
  )),
}));

vi.mock('src/components/ui', () => ({
  CharacterLimit: vi.fn(({ children, value }) => (
    <div>
      <span data-testid="limit">{value.length} / 200</span>
      {children}
    </div>
  )),
}));

function MockParenComponent() {
  const [summary, setSummary] = useState('');

  return (
    <SummaryContext.Provider value={{ summary, setSummary }}>
      <Summary />
    </SummaryContext.Provider>
  );
}

test('renders Summary', () => {
  render(<Summary />);

  expect(screen.getByTestId('form')).toMatchSnapshot();
});

test('renders initial summary values correctly', () => {
  render(<MockParenComponent />);
  expect(screen.getByLabelText('Mock label')).toHaveValue('');
  expect(screen.getByTestId('limit')).toHaveTextContent('0 / 200');
});

test('updates summary value as a user types', async () => {
  const user = userEvent.setup();

  render(<MockParenComponent />);
  const textarea = screen.getByLabelText('Mock label');
  const limitSpan = screen.getByTestId('limit');

  await user.type(textarea, 'This is a short summary');
  expect(textarea).toHaveValue('This is a short summary');
  expect(limitSpan).toHaveTextContent('23 / 200');

  await user.clear(textarea);
  await user.type(textarea, 'hello');
  expect(textarea).toHaveValue('hello');
  expect(limitSpan).toHaveTextContent('5 / 200');
});

test('renders Summary with correct number of characters when user reaches the limit', async () => {
  const user = userEvent.setup();

  render(<MockParenComponent />);
  const textarea = screen.getByLabelText('Mock label');

  await user.type(
    textarea,
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ullamcorper velit sed ullamcorper morbi. Rhoncus urna neque viverra justo nec ultrices dui sapien.',
  );

  expect(textarea.value).toHaveLength(200);
  expect(screen.getByTestId('limit')).toHaveTextContent('200 / 200');
});
