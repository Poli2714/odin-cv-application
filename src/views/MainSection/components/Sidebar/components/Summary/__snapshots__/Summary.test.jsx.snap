// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders Summary 1`] = `
<form
  action=""
  data-testid="form"
  method="get"
>
  <div>
    <label
      for="summary"
    >
      Mock label
    </label>
    <div>
      <span
        data-testid="limit"
      >
        0
         / 200
      </span>
      <textarea
        id="summary"
        maxlength="200"
        rows="10"
      />
    </div>
  </div>
</form>
`;
